src/wildcard.rs

```diff
@@ -382,10 +383,36 @@ fn wildcard_test_flags_then_complete(
         .check_type()
         .map(|x| x == DirEntryType::reg)
         .unwrap_or(false);
-    if executables_only && (!is_regular_file || waccess(filepath, X_OK) != 0) {
+    let is_executable = is_regular_file && waccess(filepath, X_OK) == 0;
+    if executables_only && !is_executable {
         return false;
     }
 
+    // For executables on Cygwin, prefer the name without the .exe, to match
+    // better with Unix names, but only if there isn't also a file without that
+    // extension and the user hasn't started to type the extension
+    let filename = if cfg!(cygwin)
+        && is_executable
+        && string_suffixes_string_case_insensitive(L!(".exe"), filename)
+        && wc.len() + 4 <= filename.len()
+    {
+        let filepath_stripped = &filepath[0..filepath.len() - 4];
+        let stat_stripped = lwstat(filepath_stripped).map(|stat| (stat.dev(), stat.ino()));
+        let stat = lwstat(filepath).map(|stat| (stat.dev(), stat.ino()));
+
+        // TODO(MSRV>=1.88) use if-let-chain
+        //   if let Ok(stat_stripped) = stat_stripped
+        //       && let Ok(stat) = stat
+        //       && stat_stripped == stat
+        if stat_stripped.is_ok() && stat.is_ok() && stat_stripped.unwrap() == stat.unwrap() {
+            &filename[0..filename.len() - 4]
+        } else {
+            filename
+        }
+    } else {
+        filename
```

Comment:
if we need to identical else branches
(presumably even with let-else),
maybe it's better to make `filename` `mut` instead?
Then we don't need any else branch.